staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/validation.go

```diff
@@ -866,6 +872,34 @@ func ValidateCustomResourceDefinitionOpenAPISchema(schema *apiextensions.JSONSch
 		}
 	}
 
+	if opts.requireMapListKeysMapSetValidation && schema.XListType != nil && (*schema.XListType == "map" || *schema.XListType == "set") && schema.Items != nil && schema.Items.Schema != nil {
+		is := schema.Items.Schema
+		if is.Nullable {
+			allErrs = append(allErrs, field.Forbidden(fldPath.Child("items").Child("schema").Child("nullable"), "entry cannot be nullable when x-kubernetes-list-type is "+*schema.XListType))
+		}
+
+		if *schema.XListType == "map" {
+			if *schema.XListType == "map" {
+				requiredMap := make(map[string]struct{})
+				for _, required := range is.Required {
+					requiredMap[required] = struct{}{}
+				}
+
+				for k, obj := range is.Properties {
+					if _, required := requiredMap[k]; !required {
+						if obj.Default == nil {
+							allErrs = append(allErrs, field.Required(fldPath.Child("items").Child("schema").Child("properties").Key(k).Child("default"), "default value must be set if key is not required"))
+						}
+					}
+
+					if obj.Nullable {
+						allErrs = append(allErrs, field.Forbidden(fldPath.Child("items").Child("schema").Child("properties").Key(k).Child("nullable"), "key cannot be nullable"))
+					}
+				}
```

Comment:
There is also `AdditionalProperties` next to explicit properties. Both are mutual exclusive in CRDs. But we could hit `AdditionalProperties` here. In that case, check the schema for those too, same conditions.